core/src/test/java/com/yahoo/oak/ComputeTest.java

```diff
@@ -116,14 +127,18 @@ public void testMain() throws InterruptedException {
             oak.zc().putIfAbsent(key, val);
         }
 
-        for (int i = 0; i < NUM_THREADS; i++) {
-            threads.get(i).start();
-        }
-
         latch.countDown();
 
-        for (int i = 0; i < NUM_THREADS; i++) {
-            threads.get(i).join();
+        executor.shutdown();
+        try {
+            executor.shutdown();
+            if (!executor.awaitTermination(10, TimeUnit.SECONDS)) {
+                executor.shutdownNow();
+                Assert.fail("should have done all the tasks in time");
+            }
+        } catch (InterruptedException e) {
+            executor.shutdownNow();
+            Assert.fail("failed to run all the tasks in the executor service");
         }
```

Comment:
I Did that by interdoucing function that will get the executor, list of pending tasks futures and timeout
and  it will wait to those futures and handle the pool closing that also deal  with the error propegation.